qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py

```diff
@@ -287,3 +703,281 @@ def use_high_resolution_adc(self) -> None:
+        self.iv_sweep.current_compliance(i_comp)
+        self.root_instrument.clear_timer_count()
+
+        error_list, error = [], ''
+
+        while error != '+0,"No Error."':
+            error = self.root_instrument.error_message()
+            error_list.append(error)
+
+        if len(error_list) <= 1:
+            self.setup_fnc_already_run = True
+        else:
+            raise Exception(f'Received following errors while trying to set '
+                            f'staircase sweep {error_list}')
+
+    def _iv_sweep_voltages(self) -> tuple:
+        sign = lambda s: s and (1, -1)[s < 0]
+        start_value = self.iv_sweep.sweep_start()
+        end_value = self.iv_sweep.sweep_end()
+        step_value = self.iv_sweep.sweep_steps()
+        sweep_mode = self.iv_sweep.sweep_mode()
+        if sweep_mode in (2, 4):
+            if not sign(start_value) == sign(end_value):
+                if sign(start_value) == 0:
+                    start_value = sign(start_value) * 0.005  # resolution
+                elif sign(end_value) == 0:
+                    end_value = sign(end_value) * 0.005  # resolution
+                else:
+                    raise AssertionError(
+                        "Polarity of start and end is not "
+                        "same.")
+
+        def linear_sweep(start: float, end: float, steps: int) -> tuple:
+            sweep_val = np.linspace(start, end, steps)
+            return tuple(sweep_val)
+
+        def log_sweep(start: float, end: float, steps: int) -> tuple:
+            sweep_val = np.logspace(np.log10(start), np.log10(end), steps)
+            return tuple(sweep_val)
+
+        def linear_2way_sweep(start: float, end: float,
+                              steps: int) -> tuple:
+            if steps % 2 == 0:
+                half_list = list(np.linspace(start, end, steps // 2))
+                sweep_val = half_list + half_list[::-1]
+            else:
+                half_list = list(np.linspace(start, end, steps // 2,
+                                             endpoint=False))
+                sweep_val = half_list + [end] + half_list[::-1]
+            return tuple(sweep_val)
+
+        def log_2way_sweep(start: float, end: float, steps: int) -> tuple:
+            if steps % 2 == 0:
+                half_list = list(
+                    np.logspace(np.log10(start), np.log10(end),
+                                steps // 2))
+                sweep_val = half_list + half_list[::-1]
+            else:
+                half_list = list(
+                    np.logspace(np.log10(start), np.log10(end),
+                                steps // 2, endpoint=False))
+                sweep_val = half_list + [end] + half_list[::-1]
+            return tuple(sweep_val)
+
+        modes = {1: linear_sweep,
+                 2: log_sweep,
+                 3: linear_2way_sweep,
+                 4: log_2way_sweep}
+
+        return modes[sweep_mode](start_value, end_value, step_value)
```

Comment:
don't forget to restructure such that this is reused for both CMU and SMU

Reply:
This is not needed in iv_staircase_sweep because we ask te source voltage from the instrument itself by using FMT1,1 format.